metagraph/src/annotation/binary_matrix/row_diff/row_diff.hpp

```diff
@@ -9,92 +9,175 @@
      */
-    std::vector<Row> get_column(Column column) const override;
+    std::vector<Row> get_column(Column column) const override  {
+        std::vector<Row> result;
+        for (Row row = 0; row < num_rows(); ++row) {
+            if (get(row, column))
+                result.push_back(row);
+        }
+        return result;
+    }
+
+    SetBitPositions get_row(Row row) const override  {
+        Vector<uint64_t> result = get_diff(row);
 
+        uint64_t boss_edge = graph_->kmer_to_boss_index(
+                graph::AnnotatedSequenceGraph::anno_to_graph_index(row));
+        const graph::boss::BOSS &boss = graph_->get_boss();
 
-    SetBitPositions get_row(Row row) const override;
+        while (!terminal_[row]) {
+            graph::boss::BOSS::TAlphabet w = boss.get_W(boss_edge);
+            assert(boss_edge > 1 && w != 0);
 
-    bool load(std::istream &f) override;
-    void serialize(std::ostream &f) const override;
+            // fwd always selects the last outgoing edge for a given node
+            boss_edge = boss.fwd(boss_edge, w % boss.alph_size);
+            row = graph::AnnotatedSequenceGraph::graph_to_anno_index(
+                    graph_->boss_to_kmer_index(boss_edge));
+            merge(&result, get_diff(row));
+        };
+        return result;
+    }
+
+    bool load(std::istream &f) override {
+        uint64_t len;
+        f.read(reinterpret_cast<char *>(&len), sizeof(uint64_t));
+        terminal_file_ = std::string(len, '\0');
+        f.read(terminal_file_.data(), len);
+        common::logger->trace("Loading terminal nodes from {}", terminal_file_);
+        std::ifstream fterm(terminal_file_, ios::binary);
+        terminal_.load(fterm);
+        fterm.close();
+
+        return diffs_.load(f);
+    }
+
+    void serialize(std::ostream &f) const override {
+        uint64_t len = terminal_file_.size();
+        f.write(reinterpret_cast<char *>(&len), sizeof(uint64_t));
+        f.write(terminal_file_.c_str(), len);
+        diffs_.serialize(f);
+    };
 
     void serialize(const std::string &name) const;
     bool load(const std::string &name);
 
     const sdsl::rrr_vector<> &terminal() const { return terminal_; }
-    const sdsl::rrr_vector<> &boundary() const { return boundary_; }
-    const sdsl::enc_vector<> &diffs() const { return diffs_; }
 
-    Vector<uint64_t> get_diff(uint64_t node_id) const;
+    const BaseMatrix &diffs() const { return diffs_; }
+    BaseMatrix &diffs() { return diffs_; }
 
-  private:
-    static void merge(Vector<uint64_t> *result, const Vector<uint64_t> &diff2);
+    Vector<uint64_t> get_diff(uint64_t node_id) const { return diffs_.get_row(node_id); }
 
-    uint64_t num_columns_ = 0;
+  private:
+    static void load_terminal(const std::string &filename, sdsl::rrr_vector<> *terminal) {
+        std::ifstream f(filename, ios::binary);
+        if (!f.good()) {
+            common::logger->error("Could not open terminal file {}", filename);
```

Comment:
it's not a file which is terminal
```suggestion
            common::logger->error("Could not open file  '{}' with anchors", filename);
```